### PR TITLE
Services_for_Orders_after_socket_connections

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: internal/services/order/order_rest.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+			CreatedAt: time.Now(),
+			UpdatedAt: res.OrderExecutionTime,
+		}
+		if err=config.DB.Table("holdings").Create(h).Error;err!=nil{
+			return err
+		}
+		if err=config.DB.Table("pending_orders").Where("order_id",res.OrderID).Delete(p).Error; err!=nil{
+			return err
+		}
+	} else if res.Status=="Half Completed" {
```
Comment (ghost) [inline internal/services/order/order_rest.go]: Half_Completed

File: internal/services/order/order_rest.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+import (
+	"Tradeasy/config"
+	"Tradeasy/internal/model"
+	"Tradeasy/internal/services/stock_exchange"
+	"time"
+)
+
+func UpdateBuyOrder(res *stock_exchange.OrderResponse) (err error) {
+	var p model.PendingOrders
+	if res.Status=="Failed"{
```
Comment (ghost) [inline internal/services/order/order_rest.go]: Status should be CAPS

File: internal/services/order/order_rest.go (hunk 1 above)
Comment (ghost) [reply internal/services/order/order_rest.go]: caps

File: internal/services/order/order_rest.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+				}
+			} else {
+				orderHist:=model.OrderHistory{
+					UserId:        check.UserId,
+					OrderId:       check.OrderId,
+					StockName:     check.StockName,
+					Quantity:      int(res.Quantity),
+					BuyPrice:      check.BuyPrice,
+					SellPrice:     int(res.AveragePrice),
+					CommissionFee: int(float64(res.Quantity*res.AveragePrice)*0.01),
```
Comment (ghost) [inline internal/services/order/order_rest.go]: keep 20

File: internal/services/order/order_rest.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+			if int(res.Quantity)>=check.Quantity{
+				res.Quantity=res.Quantity-uint(check.Quantity)
+				orderHist:=model.OrderHistory{
+					UserId:        check.UserId,
+					OrderId:       check.OrderId,
+					StockName:     check.StockName,
+					Quantity:      check.Quantity,
+					BuyPrice:      check.BuyPrice,
+					SellPrice:     int(res.AveragePrice),
+					CommissionFee: int(float64(check.Quantity*int(res.AveragePrice))*0.01),
```
Comment (ghost) [inline internal/services/order/order_rest.go]: keep 20

File: internal/services/order/order_rest.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+				if err=config.DB.Table("holdings").Where("id",check.Id).Update(check).Error;err!=nil{
+					return err
+				}
+			}
+		}
+		if err=config.DB.Table("pending_orders").Where("order_id=?",res.OrderID).Delete(p).Error;err!=nil{
+			return err
+		}
+	} else if res.Status=="Half Completed"{
```
Comment (ghost) [inline internal/services/order/order_rest.go]: half_completed

File: internal/services/order/order_rest.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+					CommissionFee: int(float64(res.Quantity*res.AveragePrice)*0.01),
+					BoughtAt:      check.OrderedAt,
+					SoldAt:        res.OrderExecutionTime,
+					CreatedAt:     time.Now(),
+				}
+				if err=config.DB.Table("order_history").Create(orderHist).Error;err!=nil{
+					return err
+				}
+				check.Quantity=check.Quantity-int(res.Quantity)
+
```
Comment (ghost) [inline internal/services/order/order_rest.go]: Balance update

**Comment by mukesh0513** [comment]
Run go fmt.

File: internal/services/order/req_res_body.go [hunk 7]
```diff
@@ -1,18 +1,23 @@
 package order
 
+const pending,completed,halfCompleted,failed,cancelled = "PENDING","COMPLETED","HALF_COMPLETED","FAILED","CANCELLED"
```
Comment (mukesh0513) [inline internal/services/order/req_res_body.go]: Keep them line separated for better readability.

File: internal/services/order/req_res_body.go [hunk 8]
```diff
@@ -1,18 +1,23 @@
...
+const buy,sell = "Buy","Sell"
```
Comment (mukesh0513) [inline internal/services/order/req_res_body.go]: Also constants name keep them as MixedCase.

File: internal/services/order/order.go [hunk 9]
```diff
@@ -1,13 +1,219 @@
...
 
-func SellOrder(BReq SellRequest) (sRes SellResponse, err error) {
-	return sRes, err
+func BuyOrder(bReq BuyRequest) (bRes stock_exchange.OrderResponse, err error) {
+	var stocks model.StocksFeed
+	var account model.TradingAccount
+
+	if bReq.BookType == market {
+		if err = config.DB.Table("stocks_feed").Where("stock_name=?", bReq.StockName).Last(&stocks).Error; err != nil {
+			return bRes, err
```
Comment (mukesh0513) [inline internal/services/order/order.go:28]: write descriptive error message all over the code.

File: internal/services/order/order.go [hunk 10]
```diff
@@ -1,13 +1,219 @@
...
+	if err = config.DB.Table("trading_account").Where("user_id=?", bReq.UserId).First(&account).Error; err != nil {
+		return bRes, err
+	}
+
+	if account.Balance < int64(bReq.Quantity*bReq.LimitPrice) {
+		return bRes, errors.New("balance is insufficient for the placed order")
+	}
+
+	account.Balance = account.Balance - int64(bReq.Quantity*bReq.LimitPrice)
+	//account.UpdatedAt=time.Now()
```
Comment (mukesh0513) [inline internal/services/order/order.go]: remove.

File: internal/services/order/order.go [hunk 11]
```diff
@@ -1,13 +1,219 @@
...
+		StockName: bReq.StockName,
+		OrderType: buy,
+		BookType:  bReq.BookType,
+		Quantity:  bReq.Quantity,
+		Status:    pending,
+	}
+
+	if bReq.BookType == market {
+		p.OrderPrice = bReq.LimitPrice
+		if err = config.DB.Table("pending_orders").Create(&p).Error; err != nil {
```
Comment (mukesh0513) [inline internal/services/order/order.go]: create can be moved out of if else as same method being used in both if else.

File: internal/services/order/order.go [hunk 12]
```diff
@@ -1,13 +1,219 @@
...
+		OrderType:  bReq.BookType,
+		LimitPrice: bReq.LimitPrice,
+	}
+
+	request, err := json.Marshal(exeOrder)
+	if err != nil {
+		return bRes, err
+	}
+	response, err := http.Post(buyOrderURL, "application/json", bytes.NewBuffer(request))
+	if err != nil {
```
Comment (mukesh0513) [inline internal/services/order/order.go:74]: Order is in pending even if exchange was not able to place your order. Move it to pending only if exchange accept your order.

**Comment by mukesh0513** [comment]
combine all var in a method and initialise at one place if feasible. It will make code more readable.

File: internal/services/order/order.go [hunk 13]
```diff
@@ -1,13 +1,219 @@
...
+		return bRes, err
+	}
+	return bRes, nil
+}
+
+func SellOrder(sReq SellRequest) (sRes stock_exchange.OrderResponse, err error) {
+	var stocks model.StocksFeed
+
+	var r result
+	if err = config.DB.Table("holdings").Select("stock_name, sum(quantity) as total_quantity").Where("user_id=? AND stock_name=?", sReq.UserId, sReq.StockName).Group("stock_name").Scan(&r).Error; err != nil {
```
Comment (mukesh0513) [inline internal/services/order/order.go]: you can breakdown this into multiple lines for better readability.

File: internal/services/order/order.go [hunk 14]
```diff
@@ -1,13 +1,219 @@
...
+	"bytes"
+	"encoding/json"
+	"errors"
+	"github.com/google/uuid"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+type result struct {
```
Comment (mukesh0513) [inline internal/services/order/order.go]: can this be renamed to be more descriptive on what is use case.

File: internal/services/order/order.go [hunk 15]
```diff
@@ -1,13 +1,219 @@
...
+	var stocks model.StocksFeed
+
+	var r result
+	if err = config.DB.Table("holdings").Select("stock_name, sum(quantity) as total_quantity").Where("user_id=? AND stock_name=?", sReq.UserId, sReq.StockName).Group("stock_name").Scan(&r).Error; err != nil {
+		return sRes, err
+	} else if r.TotalQuantity < sReq.Quantity {
+		return sRes, errors.New("sell Order quantity is higher than holdings quantity")
+	}
+
+	if sReq.BookType == "Market" {
```
Comment (mukesh0513) [inline internal/services/order/order.go]: Use contstants.

File: internal/services/order/order.go [hunk 16]
```diff
@@ -1,13 +1,219 @@
...
+	}
+	body, _ := ioutil.ReadAll(response.Body)
+	err = json.Unmarshal(body, &bRes)
+	if err != nil {
+		return bRes, err
+	}
+	return bRes, nil
+}
+
+func SellOrder(sReq SellRequest) (sRes stock_exchange.OrderResponse, err error) {
```
Comment (mukesh0513) [inline internal/services/order/order.go:99]: follow same comments of buyOrder.

File: internal/services/order/order_rest.go [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+			return err
+		}
+		p.Status = failed
+		if err = config.DB.Table("pending_orders").Where("order_id=?", res.OrderID).Updates(&p).Error; err != nil {
+			return err
+		}
+		if err = config.DB.Table("trading_account").Where("user_id=?", p.UserId).First(&account).Error; err != nil {
+			return err
+		}
+		account.Balance = account.Balance + int64(p.Quantity*p.OrderPrice)
```
Comment (mukesh0513) [inline internal/services/order/order_rest.go]: what if it was a limit order?

File: internal/services/order/order_rest.go [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+			StockName: p.StockName,
+			Quantity:  p.Quantity,
+			BuyPrice:  res.AveragePrice,
+			OrderedAt: res.OrderExecutionTime,
+		}
+		if p.BookType == market {
+			if err = config.DB.Table("trading_account").Where("user_id=?", p.UserId).First(&account).Error; err != nil {
+				return err
+			}
+			account.Balance = account.Balance + int64((p.OrderPrice-res.AveragePrice)*res.Quantity)
```
Comment (mukesh0513) [inline internal/services/order/order_rest.go]: why balance is getting added for buy order completion

File: internal/services/order/order_rest.go [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+			account.UpdatedAt = time.Now()
+			if err = config.DB.Table("trading_account").Where("user_id=?", p.UserId).Updates(&account).Error; err != nil {
+				return err
+			}
+		} else {
+			if err = config.DB.Table("trading_account").Where("user_id=?", p.UserId).First(&account).Error; err != nil {
+				return err
+			}
+			account.Balance = account.Balance + int64((p.LimitPrice-res.AveragePrice)*res.Quantity)
```
Comment (mukesh0513) [inline internal/services/order/order_rest.go]: [Comment](https://github.com/RohanDudeja/Tradeasy/pull/39/files#r685698297)

File: internal/services/order/order_rest.go [hunk 20]
```diff
@@ -0,0 +1,290 @@
...
+			StockName: p.StockName,
+			Quantity:  res.Quantity,
+			BuyPrice:  res.AveragePrice,
+			OrderedAt: res.OrderExecutionTime,
+		}
+		if p.BookType == market {
+			if err = config.DB.Table("trading_account").Where("user_id=?", p.UserId).First(&account).Error; err != nil {
+				return err
+			}
+			account.Balance = account.Balance + int64((p.OrderPrice-res.AveragePrice)*res.Quantity)
```
Comment (mukesh0513) [inline internal/services/order/order_rest.go]: [Comment](https://github.com/RohanDudeja/Tradeasy/pull/39/files#r685698297)

File: internal/services/order/order_rest.go [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+			account.UpdatedAt = time.Now()
+			if err = config.DB.Table("trading_account").Where("user_id=?", p.UserId).Updates(&account).Error; err != nil {
+				return err
+			}
+		} else {
+			if err = config.DB.Table("trading_account").Where("user_id=?", p.UserId).First(&account).Error; err != nil {
+				return err
+			}
+			account.Balance = account.Balance + int64((p.LimitPrice-res.AveragePrice)*res.Quantity)
```
Comment (mukesh0513) [inline internal/services/order/order_rest.go]: [Comment](https://github.com/RohanDudeja/Tradeasy/pull/39/files#r685698297)

File: internal/services/order/order_rest.go (hunk 18 above)
Comment (ghost) [reply internal/services/order/order_rest.go]: Its taking care of the left amount which can be added/subtracted.

File: internal/services/order/order_rest.go (hunk 19 above)
Comment (ghost) [reply internal/services/order/order_rest.go]: Its taking care of the left amount which can be added/subtracted.

File: internal/services/order/order_rest.go (hunk 20 above)
Comment (ghost) [reply internal/services/order/order_rest.go]: Its taking care of the left amount which can be added/subtracted.

File: internal/services/order/order_rest.go (hunk 21 above)
Comment (ghost) [reply internal/services/order/order_rest.go]: Its taking care of the left amount which can be added/subtracted.

File: internal/services/order/order_rest.go (hunk 17 above)
Comment (ghost) [reply internal/services/order/order_rest.go]: Added for limit order

File: internal/services/order/order.go (hunk 15 above)
Comment (ghost) [reply internal/services/order/order.go]: Done.

File: internal/services/order/order.go (hunk 16 above)
Comment (ghost) [reply internal/services/order/order.go:99]: Done.

File: internal/services/order/order.go (hunk 12 above)
Comment (ghost) [reply internal/services/order/order.go:74]: Done..moved the create part after checking

File: internal/services/order/order.go (hunk 11 above)
Comment (ghost) [reply internal/services/order/order.go]: Done

File: internal/services/order/order.go (hunk 10 above)
Comment (ghost) [reply internal/services/order/order.go]: Done.

File: internal/services/order/order.go (hunk 14 above)
Comment (ghost) [reply internal/services/order/order.go]: Changed.

File: internal/services/order/req_res_body.go (hunk 7 above)
Comment (ghost) [reply internal/services/order/req_res_body.go]: Done.

File: internal/services/order/req_res_body.go (hunk 8 above)
Comment (ghost) [reply internal/services/order/req_res_body.go]: Done.

File: internal/services/order/order.go (hunk 9 above)
Comment (ghost) [reply internal/services/order/order.go:28]: done

File: internal/services/order/order.go (hunk 13 above)
Comment (ghost) [reply internal/services/order/order.go]: Done.

File: internal/services/order/req_res_body.go [hunk 22]
```diff
@@ -1,18 +1,30 @@
 package order
 
+const Pending="PENDING"
```
Comment (mukesh0513) [inline internal/services/order/req_res_body.go]: can be combine them like 
const (
//Statuses
Completed
Partial

//Order Types
Market
Limit
)

File: internal/services/order/order.go [hunk 23]
```diff
@@ -1,13 +1,242 @@
...
+		}
+		bReq.LimitPrice = stocks.LTP
+	}
+
+	if err = config.DB.Table("trading_account").Where("user_id=?", bReq.UserId).First(&account).Error; err != nil {
+		log.Println("Error in Fetching Trading account",err)
+		return bRes, err
+	}
+
+	if account.Balance < int64(bReq.Quantity*bReq.LimitPrice) {
```
Comment (mukesh0513) [inline internal/services/order/order.go:38]: run gofmt project.

File: internal/services/order/order.go [hunk 24]
```diff
@@ -1,13 +1,242 @@
...
+		log.Println("Balance in Trading Account is not sufficient")
+		return bRes, errors.New("balance is insufficient for the placed order")
+	}
+
+	account.Balance = account.Balance - int64(bReq.Quantity*bReq.LimitPrice)
```
Comment (mukesh0513) [inline internal/services/order/order.go]: deduct money if exchange accepted order and places order in pending.

File: internal/services/order/order.go (hunk 23 above)
Comment (ghost) [reply internal/services/order/order.go:38]: done

File: internal/services/order/req_res_body.go (hunk 22 above)
Comment (ghost) [reply internal/services/order/req_res_body.go]: done

**Review by mukesh0513 — APPROVED** [review]
LGTM

**Pull request merged**